### PR TITLE
UI: Upgrade to Emotion 11

### PR DESCRIPTION
**Pull request by tooppaaa**
Issue: #13145

## What I did
Upgrade to emotion 11

## How to test

- chroma ?

### CODE REVIEW

**Comment by shilman** [comment]
@tooppaaa @ndelangen what are the implications to users? Is this a 6.2 change or a 7.0 change?

**Comment by ndelangen** [comment]
Hmm, should not impact users at all AFAIK

File: .babelrc.js [hunk 1]
```diff
@@ -61,7 +61,7 @@ module.exports = {
         '@babel/plugin-syntax-dynamic-import',
         ['@babel/plugin-proposal-class-properties', { loose: true }],
         'babel-plugin-macros',
-        ['emotion', { sourceMap: true, autoLabel: true }],
+        '@emotion',
```
Comment (ndelangen) [inline .babelrc.js:64]: is this not needed here?:
```suggestion
['@emotion', { sourceMap: true, autoLabel: 'always' }]
```

File: lib/ui/src/components/sidebar/Brand.tsx [hunk 2]
```diff
@@ -31,7 +31,7 @@ export const LogoLink = styled.a(({ theme }) => ({
   },
 }));
 
-export const Brand = withTheme(
+export const Brand = withTheme<React.ComponentType<{ theme: any }>>(
```
Comment (ndelangen) [inline lib/ui/src/components/sidebar/Brand.tsx:34]: `theme: any` ?

File: lib/ui/src/containers/menu.tsx [hunk 3]
```diff
@@ -45,7 +45,7 @@ export const useMenu = (
   showNav: boolean,
   enableShortcuts: boolean
 ) => {
-  const theme = useTheme<Theme>();
+  const theme = useTheme();
```
Comment (ndelangen) [inline lib/ui/src/containers/menu.tsx:48]: nice!

**Comment by shilman** [comment]
@ndelangen Our emotion version seems to impact users, maybe via interplay in `addon-docs`?

**Comment by shilman** [comment]
@tooppaaa any idea what's up with the test failures?

**Comment by JoshTumath** [comment]
> Hmm, should not impact users at all AFAIK

Yeah can confirm it does impact users. We use Emotion in our Storybook instance for our components, and we're prevented from upgrading our Emotion version to 11 because Storybook's Emotion code is still looking for the old package names. 😅

**Comment by shilman** [comment]
@JoshTumath I was concerned with the opposite direction. If we release this, won't it be a breaking change for users who are still on Emotion 10?

**Comment by JoshTumath** [comment]
> 
> 
> @JoshTumath I was concerned with the opposite direction. If we release this, won't it be a breaking change for users who are still on Emotion 10?

I imagine so! We got around it by setting a package alias in our Storybook WebPack config. I imagine you could do the reverse.

**Comment by mj12albert** [comment]
> > @JoshTumath I was concerned with the opposite direction. If we release this, won't it be a breaking change for users who are still on Emotion 10?
> 
> I imagine so! We got around it by setting a package alias in our Storybook WebPack config. I imagine you could do the reverse.

Would you mind sharing some details ? 🙏
I’m getting stuck at a `Can't resolve '@emotion/styled/base'` error and tried aliasing a bunch of  stuff to no effect...

**Comment by jackstevenson** [comment]
> > > @JoshTumath I was concerned with the opposite direction. If we release this, won't it be a breaking change for users who are still on Emotion 10?
> > 
> > 
> > I imagine so! We got around it by setting a package alias in our Storybook WebPack config. I imagine you could do the reverse.
> 
> Would you mind sharing some details ? 🙏
> I’m getting stuck at a `Can't resolve '@emotion/styled/base'` error and tried aliasing a bunch of stuff to no effect...

As @JoshTumath alluded too we had to alias the Emotion v10 package names to the newer v11 ones in the webpack config like so:
```js
return {
  ...config,
  resolve: {
    ...config.resolve,
    alias: {
      ...config.resolve.alias,
      '@emotion/core': join(__dirname, '..', '..', '..', '..', 'node_modules', '@emotion', 'react'),
      'emotion-theming': join(__dirname, '..', '..', '..', '..', 'node_modules', '@emotion', 'react')
    }
  }
};
```
The solution was taken from a suggestion in https://github.com/storybookjs/storybook/issues/10231 

I believe the underlying issue is there being two Emotion ThemeProvider contexts in the tree. A PR https://github.com/storybookjs/storybook/pull/10438 was raised to fix this but it's been closed as stale - it'd be great if that could be revisited as part of this Emotion 11 upgrade?

**Comment by langri-sha** [comment]
Does anyone have other suggestions on what to try if aliasing isn't fixing things? 

I'm in a monorepo with hoisted packages and Storybook is loading stories from workspaces. I've tried `require.resolve()`, hoisted the new Emotion packages all the way to the workspace root and double-checked to confirm all my packages have the expected Emotion doependencies, but still keep getting the `@emotion/styled/base` errors.

**Comment by bebraw** [comment]
@langri-sha The following worked for me (Emotion 11, Lerna monorepo):

**.storybook/main.js**

```javascript
const path = require("path");

module.exports = {
  ...
  webpackFinal: (config) => 
    ...

    const cwd = process.cwd();
    config.resolve.alias = {
      ...config.resolve.alias,
      "@emotion/core": path.join(cwd, "node_modules", "@emotion", "react"),
      "@emotion/styled": path.join(cwd, "node_modules", "@emotion", "styled"),
      "@emotion/styled-base": path.join(cwd, "node_modules", "@emotion", "styled"),
      "emotion-theming": path.join(cwd, "node_modules", "@emotion", "react"),
    };
  },
};
```

The trick was to rewrite `@emotion/styled-base` as `@emotion/styled` as hinted by the error.

**Comment by akd-io** [comment]
Hey @bebraw, thank you for your temporary solution.

Are you still pursuing this @tooppaaa? Would be awesome to have emotion 11 support out of the box. I would love to help, but I fear my skills are too weak in this area.

**Comment by sachinahya** [comment]
@bebraw Thanks, this got me most of the way there. I also had to do the same to `managerWebpack` as that contains aliases referencing the v10 packages and would give me runtime errors without it. This only became apparent to me when I cleared `node_modules/.cache/storybook`.

Also found v10 references in the default babel config which I needed to replace in order to get the CSS prop to work at runtime.

Finally I also needed to install `@emotion/styled` as a dev dependency to ensure I had v11 in `node_modules` and not the v10 that gets installed for Storybook.

In the end this setup is now working for me (Emotion 11, Storybook 6.1.18, Yarn Workspaces).

**.storybook/main.js**

```js
// Location of root node_modules
const modulesDir = path.join(__dirname, '../../../../node_modules');

const updateEmotionAliases = (config) => ({
  ...config,
  resolve: {
    ...config.resolve,
    alias: {
      ...config.resolve.alias,
      '@emotion/core': path.join(modulesDir, '@emotion/react'),
      '@emotion/styled': path.join(modulesDir, '@emotion/styled'),
      '@emotion/styled-base': path.join(modulesDir, '@emotion/styled'),
      'emotion-theming': path.join(modulesDir, '@emotion/react'),
    },
  },
});

module.exports = {
  // ...
  managerWebpack: updateEmotionAliases,
  webpackFinal: updateEmotionAliases,
  babel: (config) => {
    const getEntryIndexByName = (type, name) => {
      return config[type].findIndex((entry) => {
        const entryName = Array.isArray(entry) ? entry[0] : entry;
        return entryName.includes(name);
      });
    };

    // Replace reference to v10 of the Babel plugin to v11.
    const emotionPluginIndex = getEntryIndexByName('plugins', 'babel-plugin-emotion');
    config.plugins[emotionPluginIndex] = require.resolve('@emotion/babel-plugin');

    // Storybook's Babel config is already configured to use the new JSX runtime.
    // We just need to point it to Emotion's version.
    // https://emotion.sh/docs/css-prop#babel-preset
    const presetReactIndex = getEntryIndexByName('presets', '@babel/preset-react');
    config.presets[presetReactIndex][1].importSource = '@emotion/react';

    return config;
  },
  // ...
};
```

**Comment by visualfanatic** [comment]
I would also like to help if I can, just tell me what needs to be done and I will find some time to help.

**Comment by shilman** [comment]
Commented here: https://github.com/storybookjs/storybook/issues/13145#issuecomment-791283191

**Comment by ndelangen** [comment]
@riceboyler @tooppaaa I updated the PR, but it seems a lot of stuff is broken (still?)

```tsx
styled.pre<{}>(({ theme }) => ({}));
```

has `theme` be `any`

**Comment by ndelangen** [comment]
We'd love to have some help here @riceboyler would you want to schedule a meeting where we work on this together?

https://calendly.com/ndelangen/storybook

**Comment by osdiab** [comment]
Anything we can do to expedite this? My team is running into issues that makes Storybook unusable, and this PR looks like it's stalled out.

**Comment by shilman** [comment]
@osdiab AFAIK this is a breaking change so we can't release it until 7.0. have you tried yarn resolutions?

**Comment by osdiab** [comment]
@shilman I haven't - what specifically are you suggesting?

**Comment by irudoy** [comment]
@osdiab what problem are you facing? I recently had success upgrading emotion to v11 with storybook. Not without weird hacks, of course 😄

**Comment by corbt** [comment]
> @osdiab what problem are you facing? I recently had success upgrading emotion to v11 with storybook. Not without weird hacks, of course 😄

FWIW, I ran into this issue when using Storybook within a [Redwood](https://redwoodjs.com/) project, which doesn't give me enough low-level access to the Storybook config files to perform the hacks recommended in this thread. Had to abandon Storybook for now as a result. Just a motivating example of where the hacky workaround doesn't always work. :)

**Comment by osdiab** [comment]
Right now I'm not sure the root cause but when I use the css prop with a style I've defined, that works fine on my actual website, it causes Storybook to apparently infinite loop and crash in my browser. Giving up on trying to make it work for now unless someone has a suggestion on what might be going on lol happy to provide a code sample when I'm in front of my computer.

**Comment by levymetal** [comment]
Thanks for the tip @karol-f. We had to do one additional thing to get it working; we had to add `@emotion/sheet@^1.0.3` to our package. Without this, `@emotion/sheet` gets hoisted incorrectly; `@emotion/sheet@0.9.4` from `@storybook#theming#@emotion#core#@emotion#cache` gets hoisted to `@emotion/sheet`, and Emotion 11 pulls that instead of `@emotion/sheet@^1.0.3` throwing a `TypeError: cache.sheet.hydrate is not a function`. Manually adding `@emotion/sheet@^1.0.3` to the package hoists it to `@emotion/sheet`. Also, this only seems to happen with yarn and not npm.

This looks like an issue with yarn or Emotion, not Storybook. Tracking it on a similar (closed) issue: https://github.com/emotion-js/emotion/issues/1974#issuecomment-962703651. Repro here: https://github.com/levymetal/emotion-11-conflict-repro

**Comment by Andarist** [comment]
@levymetal you have misconfigured your webpack's `config.resolve.modules`. Try this patch and see that it works then:
```diff
diff --git a/webpack.config.js b/webpack.config.js
index 74479f3..b2ba83a 100644
--- a/webpack.config.js
+++ b/webpack.config.js
@@ -3,7 +3,7 @@ const path = require("path");
 
 const paths = {
   src: path.join(__dirname, "src"),
-  nodeModules: path.join(__dirname, "node_modules"),
+  nodeModules: path.join(".", "node_modules"),
 };
 
 module.exports = {
@@ -32,7 +32,7 @@ module.exports = {
       {
         test: /\.(js|jsx)$/,
         include: [paths.src],
-        exclude: [paths.nodeModules],
+        exclude: /\/node_modules\//,
         use: ["babel-loader"],
       },
     ],
```

See in the webpack docs how relative and absolute paths are treated differently in this context: https://webpack.js.org/configuration/resolve/#resolvemodules . Basically, you have made webpack to load `@emotion/sheet` from your root `node_modules` - that's why it was loading it "incorrectly".

**Comment by levymetal** [comment]
@Andarist I can't thank you enough for your response. That resolves the issue not just in the repro, but also the actual project I'm working on. Our webpack config has been misconfigured like this for over 5 years 💀. Now many of the other issues we've had suddenly make sense.

**Comment by dantman** [comment]
This is blocked till the next major, not inactive. And it's a PR with actual code contributed, not an issue. It shouldn't be free for a bot to carelessly close.

> We are always open to contributions so please send us a pull request if you would like to help. Inactive issues will be closed after 30 days. Thanks!

This message looks poorly written for whatever rule is triggering it. The bot is telling users it's going to close a PR and they should contribute a PR instead.

**Comment by jorisw** [comment]
I've always thought stalebots are a stupid thing.

Just because an issue was sufficiently documented and therefore received no new messages, it's somehow no longer relevant?

And this after only 30 days?

**Comment by mrlubos** [comment]
> @bebraw Thanks, this got me most of the way there. I also had to do the same to `managerWebpack` as that contains aliases referencing the v10 packages and would give me runtime errors without it. This only became apparent to me when I cleared `node_modules/.cache/storybook`.
> 
> Also found v10 references in the default babel config which I needed to replace in order to get the CSS prop to work at runtime.
> 
> Finally I also needed to install `@emotion/styled` as a dev dependency to ensure I had v11 in `node_modules` and not the v10 that gets installed for Storybook.
> 
> In the end this setup is now working for me (Emotion 11, Storybook 6.1.18, Yarn Workspaces).
> 
> **.storybook/main.js**
> 
> ```js
> // Location of root node_modules
> const modulesDir = path.join(__dirname, '../../../../node_modules');
> 
> const updateEmotionAliases = (config) => ({
>   ...config,
>   resolve: {
>     ...config.resolve,
>     alias: {
>       ...config.resolve.alias,
>       '@emotion/core': path.join(modulesDir, '@emotion/react'),
>       '@emotion/styled': path.join(modulesDir, '@emotion/styled'),
>       '@emotion/styled-base': path.join(modulesDir, '@emotion/styled'),
>       'emotion-theming': path.join(modulesDir, '@emotion/react'),
>     },
>   },
> });
> 
> module.exports = {
>   // ...
>   managerWebpack: updateEmotionAliases,
>   webpackFinal: updateEmotionAliases,
>   babel: (config) => {
>     const getEntryIndexByName = (type, name) => {
>       return config[type].findIndex((entry) => {
>         const entryName = Array.isArray(entry) ? entry[0] : entry;
>         return entryName.includes(name);
>       });
>     };
> 
>     // Replace reference to v10 of the Babel plugin to v11.
>     const emotionPluginIndex = getEntryIndexByName('plugins', 'babel-plugin-emotion');
>     config.plugins[emotionPluginIndex] = require.resolve('@emotion/babel-plugin');
> 
>     // Storybook's Babel config is already configured to use the new JSX runtime.
>     // We just need to point it to Emotion's version.
>     // https://emotion.sh/docs/css-prop#babel-preset
>     const presetReactIndex = getEntryIndexByName('presets', '@babel/preset-react');
>     config.presets[presetReactIndex][1].importSource = '@emotion/react';
> 
>     return config;
>   },
>   // ...
> };
> ```

Hey everyone! I used to use this excellent quote trick, but found that it's no longer necessary after disabling [`emotionAlias`](https://github.com/storybookjs/storybook/pull/13300#issuecomment-783268111) like this in your `main.js`:

```
features: {
  emotionAlias: false
}
```

**Comment by ronaldruzicka** [comment]
@mrlubos thanks a lot! I spent almost 2 days figuring this thing out.

**Comment by Andarist** [comment]
This is being superseded by https://github.com/storybookjs/storybook/pull/17640

**Pull request closed without merging**